src/transformers/models/wav2vec2/tokenization_wav2vec2.py

```diff
@@ -204,31 +224,110 @@ def _convert_id_to_token(self, index: int) -> str:
-            join_token = " "
-        else:
-            join_token = ""
-
         # replace delimiter token
-        string = join_token.join(
-            [" " if token == self.word_delimiter_token else token for token in filtered_tokens]
-        ).strip()
+        processed_tokens = [" " if token == self.word_delimiter_token else token for token in filtered_tokens]
+
+        time_stamps = word_time_stamps = None
+        if output_time_stamps:
+            if stride is None:
+                raise ValueError(
+                    "Make sure to provide the `stride` of the audio model "
+                    "in order to retrieve the time stamps for each token."
+                    "The stride is usually defined as `model.stride`."
+                )
+
+            if sampling_rate is None:
+                raise ValueError(
+                    "Make sure to provide the `sampling_rate` of the feature extractor "
+                    "in order to retrieve the time stamps for each token."
+                    "The sampling rate is usually defined as `feature_extractor.sampling_rate`."
+                )
+
+            frame_context_in_seconds = stride / sampling_rate
+
+            end_indices = np.asarray(token_repetitions).cumsum()
+            start_indices = np.concatenate(([0], end_indices[:-1]))
+
+            # create time stamps
+            time_stamps = [
+                {"token": t, "start_time": s * frame_context_in_seconds, "end_time": e * frame_context_in_seconds}
+                for t, s, e in zip(tokens, start_indices, end_indices)
+            ]
+
+            # filter out CTC token
+            time_stamps = list(filter(lambda time_stamp: time_stamp["token"] != self.pad_token, time_stamps))
+
+            if len(time_stamps) != len(processed_tokens):
+                raise ValueError(
+                    f"`time_stamps`: {time_stamps} and `processed_tokens`: {processed_tokens}"
+                    " have to be of the same length, but are: "
+                    f"`len(time_stamps)`: {len(time_stamps)} and `len(processed_tokens)`:"
+                    f" {len(processed_tokens)}"
+                )
+
+            # set tokens to correct
+            for i, token in enumerate(processed_tokens):
+                time_stamps[i]["token"] = token
+
+            # retrieve word time stamps from time stamps
+            word_time_stamps = self._get_word_time_stamps(time_stamps)
+
+        # join to string
+        join_char = " " if spaces_between_special_tokens else ""
+        string = join_char.join(processed_tokens).strip()
 
         if self.do_lower_case:
             string = string.lower()
-        return string
+
+        return {
+            "text": string,
+            "token_time_stamps": time_stamps,
+            "word_time_stamps": word_time_stamps,
+        }
+
+    @staticmethod
+    def _get_word_time_stamps(time_stamps: Dict[str, Union[str, float]]) -> Dict[str, Union[str, float]]:
```

Comment:
I don't think we should attempt to do "word" boundaries directly here (maybe a helper function).

First we're output roughly twice the same data, then, word don't mean whitespace separated in many languages (like chinese, I even argue it's the case in German but it's not the point).

Taking only `subtitling` for instance as a use case, you don't really want "word" but you want "sentences", with a cap on length of text. You may even get sentence boundary, not looking at text, but looking at the padding length in seconds between spoken text (`token_timestamp[i + 1].stop - token_timestamp[i].start`, no idea if it works, but definitely seems like a sound strategy for subtitling I would like to attempt).

Making it *raw* by default is good, (pure timestamps per character). Making it easy to retrieve custom boundaries simple would be greater. Using your function as an option (in the docs maybe) seems better as opt-in.

Reply:
I understand your point, but disagree here a bit because the tokenizer also outputs text as words - it has a specific `word_delimiter_token` and clearly puts a white space between words which holds true for almost all languages (except the few ones that are character based)

Reply:
It makes sense. I'll trust your judgement here.

I do tend to push back on word everywhere I can, but you are correct, in wav2vec2 there is a core concept of word_delimiter_token